pkg/controller/apimanager/apimanager_controller.go

```diff
@@ -151,111 +154,139 @@ func (r *ReconcileAPIManager) Reconcile(request reconcile.Request) (reconcile.Re
-				if err != nil {
-					reqLogger.Error(err, fmt.Sprintf("Error creating object %s. Requeuing request...", objectInfo))
-					return reconcile.Result{}, err
+				createErr := r.client.Create(context.TODO(), obj)
+				if createErr != nil {
+					r.reqLogger.Error(createErr, fmt.Sprintf("Error creating object %s. Requeuing request...", objectInfo))
+					return reconcile.Result{}, createErr
 				}
-				reqLogger.Info(fmt.Sprintf("Created object %s", objectInfo))
+				r.reqLogger.Info(fmt.Sprintf("Created object %s", objectInfo))
 			} else {
-				reqLogger.Error(err, fmt.Sprintf("Failed to get %s.  Requeuing request...", objectInfo))
+				r.reqLogger.Error(err, fmt.Sprintf("Failed to get %s.  Requeuing request...", objectInfo))
 				return reconcile.Result{}, err
 			}
 		} else {
-			reqLogger.Info(fmt.Sprintf("Object %s already exists", objectInfo))
-		}
-
-		// Update secrets with consistent data
-		if secret, ok := objCopy.(*v1.Secret); ok {
-			reqLogger.Info("Object is a secret. Updating object...")
-			err = r.client.Update(context.TODO(), secret)
-			if err != nil {
-				reqLogger.Error(err, fmt.Sprintf("Failed to update secret secret/%s. Requeuing request...", secret.Name))
-				return reconcile.Result{}, err
+			r.reqLogger.Info(fmt.Sprintf("Object %s already exists", objectInfo))
+			if secret, ok := objCopy.(*v1.Secret); ok {
+				r.reqLogger.Info(fmt.Sprintf("Object %s is a secret. Reconciling it...", objectInfo))
+				// We get copy to avoid modifying possibly obtained object
+				// from the cache
+				foundSecret := found.(*v1.Secret)
+				r.updateSecret(secret, foundSecret, instance)
+				if err != nil {
+					r.reqLogger.Error(err, fmt.Sprintf("Failed to update secret secret/%s. Requeuing request...", secret.Name))
+					return reconcile.Result{}, err
+				}
 			}
 		}
-		// Here means that the object has been able to be obtained
-		// and checking for differences should be done to reconcile possible
-		// differences that we want to handle
 	}
 
-	reqLogger.Info("Finished Current reconcile request successfully. Skipping requeue of the request")
+	r.reqLogger.Info("Finished Current reconcile request successfully. Skipping requeue of the request")
 	return reconcile.Result{}, nil
 }
 
-func createAPIManager(cr *appsv1alpha1.APIManager, client client.Client) ([]runtime.RawExtension, error) {
-	results, err := createAPIManagerObjects(cr, client)
+func (r *ReconcileAPIManager) updateSecret(desired, current *v1.Secret, cr *appsv1alpha1.APIManager) error {
+	// We copy the secrets because we don't know the source of them. Might
+	// come from the Cache
+	currentCopy := current.DeepCopy()
+	desiredCopy := desired.DeepCopy()
+
+	// We convert StringData to Data because stringData cannot be read when
+	// obtained from the Kubernetes API and we need to compare the secret
+	// data
+	desiredCopy.Data = secretStringDataToData(desiredCopy.StringData)
+	if secretsEqual(currentCopy, desiredCopy) {
```

Comment:
`updateSecret` method name makes me think you are updating secrets. Why not check if secrets are equal and if not call `updateSecret`?

Reply:
That's what I'm doing later in https://github.com/3scale/3scale-operator/blob/834e7b6b050731578d08bbda4fd443c954f39168/pkg/controller/apimanager/apimanager_controller.go#L211

Reply:
Maybe renaming the "updateSecret" method to "reconcileSecret" ?

Reply:
I renamed the method updateSecret to reconcileSecret to make more clear that not always an update will be produced